File: server/src/services/original_text_service.ts
import { PrismaClient, Prisma } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';
import dedup from '@/utils/dedup';

const prisma = new PrismaClient();

interface ListParams {
    keyword?: string;
    categoryType?: string;
    status?: number;
    auditStatus?: number;
    page?: number;
    pageSize?: number;
}

interface OriginalTextSummary {
    id: string;
    title: string;
    author?: string;
    source?: string;
    categoryType?: string;
    summary?: string;
    wordCount: number;
    status: number;
    auditStatus?: number;
    quoteCount: number;
    createdAt: string;
    updatedAt: string;
}

interface OriginalTextDetail extends OriginalTextSummary {
    content: string;
    pinyinData?: string;
    sourceUrl?: string;
    tags?: string;
    provider?: string;
    complianceTag?: string;
    protocol?: string;
    quotes: Array<{
        id: string;
        content: string;
        author?: string;
        source?: string;
    }>;
}

interface PageResult<T> {
    total: number;
    list: T[];
}

function countWords(text: string): number {
    if (!text) return 0;
    const cleaned = text.replace(/\s+/g, '');
    return cleaned.length;
}

async function list(params: ListParams): Promise<PageResult<OriginalTextSummary>> {
    const page = params.page ?? 1;
    const pageSize = params.pageSize ?? 20;
    const skip = (page - 1) * pageSize;

    const where: Prisma.OriginalTextWhereInput = {};

    if (params.auditStatus !== undefined && params.auditStatus !== null && !isNaN(params.auditStatus)) {
        where.auditStatus = params.auditStatus;
    }

    if (params.keyword) {
        where.OR = [
            { title: { contains: params.keyword } },
            { content: { contains: params.keyword } },
            { author: { contains: params.keyword } },
        ];
    }

    if (params.categoryType) {
        where.categoryType = params.categoryType;
    }

    if (params.status !== undefined && params.status !== null) {
        where.status = params.status;
    }

    const [total, records] = await Promise.all([
        prisma.originalText.count({ where }),
        prisma.originalText.findMany({
            where,
            skip,
            take: pageSize,
            orderBy: { id: 'desc' },
        }),
    ]);

    const list: OriginalTextSummary[] = records.map((r) => ({
        id: String(r.id),
        title: r.title,
        author: r.author ?? undefined,
        source: r.source ?? undefined,
        categoryType: r.categoryType ?? undefined,
        summary: r.summary ?? undefined,
        wordCount: r.wordCount,
        status: r.status,
        auditStatus: r.auditStatus ?? undefined,
        quoteCount: r.quoteCount,
        createdAt: r.createdAt.toISOString(),
        updatedAt: r.updatedAt.toISOString(),
    }));

    return { total, list };
}

async function getById(id: number, auditStatus?: number): Promise<OriginalTextDetail | null> {
    const where: Prisma.OriginalTextWhereInput = { id };
    if (auditStatus !== undefined && auditStatus !== null && !isNaN(auditStatus)) {
        where.auditStatus = auditStatus;
    }
    const record = await prisma.originalText.findUnique({
        where,
        include: {
            quotes: {
                select: {
                    id: true,
                    content: true,
                    author: true,
                    source: true,
                },
                orderBy: { id: 'asc' },
            },
        },
    });

    if (!record) return null;

    return {
        id: String(record.id),
        title: record.title,
        author: record.author ?? undefined,
        source: record.source ?? undefined,
        sourceUrl: record.sourceUrl ?? undefined,
        categoryType: record.categoryType ?? undefined,
        content: record.content,
        pinyinData: record.pinyinData ?? undefined,
        summary: record.summary ?? undefined,
        tags: record.tags ?? undefined,
        provider: record.provider ?? undefined,
        complianceTag: record.complianceTag ?? undefined,
        protocol: record.protocol ?? undefined,
        wordCount: record.wordCount,
        status: record.status,
        auditStatus: record.auditStatus ?? undefined,
        quoteCount: record.quoteCount,
        createdAt: record.createdAt.toISOString(),
        updatedAt: record.updatedAt.toISOString(),
        quotes: record.quotes.map((q) => ({
            id: String(q.id),
            content: q.content,
            author: q.author ?? undefined,
            source: q.source ?? undefined,
        })),
    };
}

async function create(data: {
    title: string;
    content: string;
    author?: string;
    source?: string;
    sourceUrl?: string;
    categoryType?: string;
    summary?: string;
    tags?: string;
    provider?: string;
    complianceTag?: string;
    protocol?: string;
    status?: number;
}): Promise<OriginalTextDetail> {
    const record = await prisma.originalText.create({
        data: {
            title: data.title,
            content: data.content,
            author: data.author ?? null,
            source: data.source ?? null,
            sourceUrl: data.sourceUrl ?? null,
            categoryType: data.categoryType ?? null,
            summary: data.summary ?? null,
            tags: data.tags ?? null,
            provider: data.provider ?? null,
            complianceTag: data.complianceTag ?? null,
            protocol: data.protocol ?? null,
            wordCount: countWords(data.content),
            status: data.status ?? 1,
        },
    });

    const detail = await getById(record.id);
    if (!detail) throw new BusinessError(500, '创建后查询失败');
    return detail;
}

async function update(
    id: number,
    data: {
        title?: string;
        content?: string;
        author?: string;
        source?: string;
        sourceUrl?: string;
        categoryType?: string;
        summary?: string;
        tags?: string;
        provider?: string;
        complianceTag?: string;
        protocol?: string;
        status?: number;
    }
): Promise<OriginalTextDetail> {
    const existing = await prisma.originalText.findUnique({ where: { id } });
    if (!existing) throw new BusinessError(404, '原文不存在');

    const updateData: Prisma.OriginalTextUpdateInput = {
        title: data.title,
        author: data.author !== undefined ? data.author : undefined,
        source: data.source !== undefined ? data.source : undefined,
        sourceUrl: data.sourceUrl !== undefined ? data.sourceUrl : undefined,
        categoryType: data.categoryType !== undefined ? data.categoryType : undefined,
        summary: data.summary !== undefined ? data.summary : undefined,
        tags: data.tags !== undefined ? data.tags : undefined,
        provider: data.provider !== undefined ? data.provider : undefined,
        complianceTag: data.complianceTag !== undefined ? data.complianceTag : undefined,
        protocol: data.protocol !== undefined ? data.protocol : undefined,
        status: data.status,
    };

    if (data.content !== undefined) {
        updateData.content = data.content;
        updateData.wordCount = countWords(data.content);
    }

    await prisma.originalText.update({ where: { id }, data: updateData });

    const detail = await getById(id);
    if (!detail) throw new BusinessError(500, '更新后查询失败');
    return detail;
}

async function remove(id: number): Promise<void> {
    const existing = await prisma.originalText.findUnique({ where: { id } });
    if (!existing) throw new BusinessError(404, '原文不存在');

    await prisma.$transaction(async (tx) => {
        await tx.quote.deleteMany({ where: { originalTextId: id } });
        await tx.originalText.delete({ where: { id } });
    });
}

async function addQuoteFromText(
    originalTextId: number,
    quoteContent: string,
    author?: string,
    source?: string
): Promise<{ id: string; content: string; author?: string; source?: string }> {
    const originalText = await prisma.originalText.findUnique({ where: { id: originalTextId } });
    if (!originalText) throw new BusinessError(404, '原文不存在');

    const trimmed = quoteContent.trim();
    if (!trimmed) throw new BusinessError(400, '金句内容不能为空');

    // 校验选段是否真的来自原文（允许前后各 5 字的宽松范围，因为清洗时可能有空白差异）
    if (!originalText.content.includes(trimmed)) {
        throw new BusinessError(400, '选段内容与原文不匹配，请检查是否从原文中选取');
    }

    const fp = dedup.fingerprint(trimmed);

    // 系统添加的金句要 MD5 去重
    const existingDup = await prisma.quote.findFirst({
        where: { contentMd5: fp.contentMd5 },
    });
    if (existingDup) {
        throw new BusinessError(409, `该金句已存在（#${existingDup.id}）`);
    }

    const quote = await prisma.$transaction(async (tx) => {
        const created = await tx.quote.create({
            data: {
                content: trimmed,
                contentMd5: fp.contentMd5,
                simHash: fp.simHash,
                author: author ?? originalText.author,
                source: source ?? originalText.source ?? originalText.title,
                sourceUrl: originalText.sourceUrl,
                originalTextId,
                wordCount: countWords(trimmed),
                isFree: true,
                isActive: true,
                provider: originalText.provider ?? 'admin',
                quoteType: 'admin',
                aiExtracted: false,
            },
        });

        await tx.originalText.update({
            where: { id: originalTextId },
            data: { quoteCount: { increment: 1 } },
        });

        return created;
    });

    return {
        id: String(quote.id),
        content: quote.content,
        author: quote.author ?? undefined,
        source: quote.source ?? undefined,
    };
}

export { list, getById, create, update, remove, addQuoteFromText };

File: server/app/services/original_text_service.py
"""原文服务

对应 TypeScript 版 server/src/services/original_text_service.ts
"""
from datetime import datetime

from sqlalchemy import or_
from sqlalchemy.orm import Session

from ..deps import BusinessError
from ..models import OriginalText, Quote
from ..utils.dedup import dedup


def _count_words(text: str) -> int:
    if not text:
        return 0
    import re
    cleaned = re.sub(r"\s+", "", text)
    return len(cleaned)


def _serialize_summary(r: OriginalText) -> dict:
    return {
        "id": str(r.id),
        "title": r.title,
        "author": r.author or None,
        "source": r.source or None,
        "categoryType": r.categoryType or None,
        "summary": r.summary or None,
        "wordCount": r.wordCount,
        "status": r.status,
        "auditStatus": r.auditStatus or None,
        "quoteCount": r.quoteCount,
        "createdAt": r.createdAt.isoformat() if r.createdAt else None,
        "updatedAt": r.updatedAt.isoformat() if r.updatedAt else None,
    }


def _serialize_detail(db: Session, record: OriginalText) -> dict:
    quotes = (
        db.query(Quote)
        .filter(Quote.originalTextId == record.id)
        .order_by(Quote.id.asc())
        .all()
    )
    return {
        "id": str(record.id),
        "title": record.title,
        "author": record.author or None,
        "source": record.source or None,
        "sourceUrl": record.sourceUrl or None,
        "categoryType": record.categoryType or None,
        "content": record.content,
        "pinyinData": record.pinyinData or None,
        "summary": record.summary or None,
        "tags": record.tags or None,
        "provider": record.provider or None,
        "complianceTag": record.complianceTag or None,
        "protocol": record.protocol or None,
        "wordCount": record.wordCount,
        "status": record.status,
        "auditStatus": record.auditStatus or None,
        "quoteCount": record.quoteCount,
        "createdAt": record.createdAt.isoformat() if record.createdAt else None,
        "updatedAt": record.updatedAt.isoformat() if record.updatedAt else None,
        "quotes": [
            {
                "id": str(q.id),
                "content": q.content,
                "author": q.author or None,
                "source": q.source or None,
            }
            for q in quotes
        ],
    }


def list(db: Session, params: dict | None = None) -> dict:
    if params is None:
        params = {}
    page = params.get("page") or 1
    page_size = params.get("pageSize") or 20
    skip = (page - 1) * page_size

    conditions = []

    audit_status = params.get("auditStatus")
    if audit_status is not None:
        conditions.append(OriginalText.auditStatus == audit_status)

    keyword = params.get("keyword")
    if keyword:
        conditions.append(
            or_(
                OriginalText.title.contains(keyword),
                OriginalText.content.contains(keyword),
                OriginalText.author.contains(keyword),
            )
        )

    category_type = params.get("categoryType")
    if category_type:
        conditions.append(OriginalText.categoryType == category_type)

    status = params.get("status")
    if status is not None:
        conditions.append(OriginalText.status == status)

    from sqlalchemy import func

    query = db.query(OriginalText)
    if conditions:
        query = query.filter(*conditions)

    total = query.count()
    records = (
        query
        .order_by(OriginalText.id.asc())
        .offset(skip)
        .limit(page_size)
        .all()
    )

    return {"total": total, "list": [_serialize_summary(r) for r in records]}


def getById(db: Session, id: int, auditStatus: int | None = None) -> dict | None:
    query = db.query(OriginalText).filter(OriginalText.id == id)
    if auditStatus is not None:
        query = query.filter(OriginalText.auditStatus == auditStatus)

    record = query.first()
    if not record:
        return None

    return _serialize_detail(db, record)


def create(db: Session, data: dict) -> dict:
    record = OriginalText(
        title=data["title"],
        content=data["content"],
        author=data.get("author"),
        source=data.get("source"),
        sourceUrl=data.get("sourceUrl"),
        categoryType=data.get("categoryType"),
        summary=data.get("summary"),
        tags=data.get("tags"),
        provider=data.get("provider"),
        complianceTag=data.get("complianceTag"),
        protocol=data.get("protocol"),
        wordCount=_count_words(data["content"]),
        status=data.get("status", 1),
    )
    db.add(record)
    db.commit()
    db.refresh(record)

    detail = getById(db, record.id)
    if not detail:
        raise BusinessError(500, "创建后查询失败", 500)
    return detail


def update(db: Session, id: int, data: dict) -> dict:
    existing = db.query(OriginalText).filter(OriginalText.id == id).first()
    if not existing:
        raise BusinessError(404, "原文不存在", 404)

    for field in ["title", "author", "source", "sourceUrl", "categoryType",
                   "summary", "tags", "provider", "complianceTag", "protocol", "status"]:
        if field in data:
            setattr(existing, field, data[field])

    if "content" in data:
        existing.content = data["content"]
        existing.wordCount = _count_words(data["content"])

    existing.updatedAt = datetime.now()
    db.commit()
    db.refresh(existing)

    detail = getById(db, id)
    if not detail:
        raise BusinessError(500, "更新后查询失败", 500)
    return detail


def remove(db: Session, id: int) -> None:
    existing = db.query(OriginalText).filter(OriginalText.id == id).first()
    if not existing:
        raise BusinessError(404, "原文不存在", 404)

    try:
        db.query(Quote).filter(Quote.originalTextId == id).delete()
        db.query(OriginalText).filter(OriginalText.id == id).delete()
        db.commit()
    except Exception:
        db.rollback()
        raise


def addQuoteFromText(
    db: Session,
    originalTextId: int,
    quoteContent: str,
    author: str | None = None,
    source: str | None = None,
) -> dict:
    original_text = db.query(OriginalText).filter(OriginalText.id == originalTextId).first()
    if not original_text:
        raise BusinessError(404, "原文不存在", 404)

    trimmed = quoteContent.strip()
    if not trimmed:
        raise BusinessError(400, "金句内容不能为空")

    if trimmed not in original_text.content:
        raise BusinessError(400, "选段内容与原文不匹配，请检查是否从原文中选取")

    fp = dedup["fingerprint"](trimmed)

    existing_dup = db.query(Quote).filter(Quote.contentMd5 == fp["contentMd5"]).first()
    if existing_dup:
        raise BusinessError(409, f"该金句已存在（#{existing_dup.id}）")

    try:
        quote = Quote(
            content=trimmed,
            contentMd5=fp["contentMd5"],
            simHash=fp["simHash"],
            author=author or original_text.author,
            source=source or original_text.source or original_text.title,
            sourceUrl=original_text.sourceUrl,
            originalTextId=originalTextId,
            wordCount=_count_words(trimmed),
            isFree=True,
            isActive=True,
            provider=original_text.provider or "admin",
            quoteType="admin",
            aiExtracted=False,
        )
        db.add(quote)

        original_text.quoteCount = (original_text.quoteCount or 0) + 1
        original_text.updatedAt = datetime.now()

        db.commit()
        db.refresh(quote)
    except Exception:
        db.rollback()
        raise

    return {
        "id": str(quote.id),
        "content": quote.content,
        "author": quote.author or None,
        "source": quote.source or None,
    }
